server/proxy/proxy.go

```diff
@@ -37,75 +37,61 @@ func (p *Proxy) Call(sessionID, actingUserID string, creq *apps.CallRequest) *ap
 		BotUsername: app.BotUsername,
 	}
 
-	cresp := p.callApp(app, sessionID, actingUserID, creq)
+	cresp := p.callApp(in, app, creq)
 	return apps.NewProxyCallResponse(cresp, metadata)
 }
 
-func (p *Proxy) callApp(app *apps.App, sessionID, actingUserID string, creq *apps.CallRequest) *apps.CallResponse {
-	if !p.AppIsEnabled(app) {
+func (p *Proxy) callApp(in Incoming, app *apps.App, creq apps.CallRequest) apps.CallResponse {
+	if !p.appIsEnabled(app) {
 		return apps.NewErrorCallResponse(errors.Errorf("%s is disabled", app.AppID))
 	}
 
-	if actingUserID != "" {
-		creq.Context.ActingUserID = actingUserID
-		creq.Context.UserID = actingUserID
-	}
-
 	if creq.Path[0] != '/' {
 		return apps.NewErrorCallResponse(utils.NewInvalidError("call path must start with a %q: %q", "/", creq.Path))
 	}
-
 	cleanPath, err := utils.CleanPath(creq.Path)
 	if err != nil {
 		return apps.NewErrorCallResponse(err)
 	}
 	creq.Path = cleanPath
 
-	up, err := p.upstreamForApp(&app.Manifest)
+	up, err := p.upstreamForApp(app)
 	if err != nil {
 		return apps.NewErrorCallResponse(err)
 	}
 
-	// Clear any ExpandedContext as it should always be set by an expander for security reasons
-	creq.Context.ExpandedContext = apps.ExpandedContext{}
-
-	conf, _, log := p.conf.Basic()
-	cc := conf.SetContextDefaultsForApp(creq.Context.AppID, creq.Context)
-
-	expander := p.newExpander(cc, p.conf, p.store, sessionID)
-	cc, err = expander.ExpandForApp(app, creq.Expand)
+	cc := creq.Context
+	cc = in.updateContext(cc)
+	creq.Context, err = p.expandContext(in, app, &cc, creq.Expand)
 	if err != nil {
 		return apps.NewErrorCallResponse(err)
 	}
-	clone := *creq
-	clone.Context = cc
 
-	callResponse := upstream.Call(up, app, &clone)
-
-	if callResponse.Type == "" {
-		callResponse.Type = apps.CallResponseTypeOK
+	cresp := upstream.Call(up, *app, creq)
+	if cresp.Type == "" {
+		cresp.Type = apps.CallResponseTypeOK
 	}
 
-	if callResponse.Form != nil {
-		if callResponse.Form.Icon != "" {
-			icon, err := normalizeStaticPath(conf, cc.AppID, callResponse.Form.Icon)
+	if cresp.Form != nil {
+		if cresp.Form.Icon != "" {
+			conf, _, log := p.conf.Basic()
+			log = log.With("app_id", app.AppID)
```

Comment:
Nit pick: The idiomatic style in the code base is to assign `log` and `conf` at the start of the function.